### PR TITLE
H-2615: Refactor `harpc/wire-protocol` to make use of `Buf/BufMut` instead of `AsyncRead/AsyncWrite`

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

This adjusts the `harpc/wire-protocol` to make use of synchronous code, expecting that all data is present, instead of using `AsyncRead/AsyncWrite`, this has been decided internally, as it provides greater guarantees, throughput as well as enables the use of `tokio-util` to convert between streams more easily.

Most of this diff are simply code changes to accommodate the new paradigm.

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: libs/@local/codec/src/harpc/wire.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    fn encode(&mut self, item: T, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let mut buffer = Buffer::new(dst);
+
+        item.encode(&mut buffer)
+            .change_context(io::Error::from(io::ErrorKind::InvalidData))
+    }
+}
+
+pub struct ProtocolDecoder<T>(PhantomData<fn() -> *const T>);
```
Comment (TimDiekmann) [inline libs/@local/codec/src/harpc/wire.rs:32]: The type of `ProtocolEncoder` and `ProtocolDecoder` is the same. Can't we reuse it?
```suggestion
pub struct ProtocolCodec<T>(PhantomData<fn() -> *const T>);
```

File: libs/@local/harpc/wire-protocol/src/codec/buffer.rs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+            return Err(Report::new(BufferError::EarlyEndOfStream));
+        }
+
+        let mut bytes = [0; N];
+        self.0.copy_to_slice(&mut bytes);
+
+        Ok(bytes)
+    }
+
+    pub(crate) fn next_discard(&mut self, count: usize) -> Result<(), BufferError> {
```
Comment (TimDiekmann) [inline libs/@local/harpc/wire-protocol/src/codec/buffer.rs:119]: As we don't return anything I think I would rename it to `discard`

**Review by TimDiekmann — APPROVED** [review]
I think this is the right choice to use `sync` over `async` here. Thank you very much for the work 🎉 
I have some non-blocking suggestions.

File: libs/@local/harpc/wire-protocol/src/codec/buffer.rs (hunk 2 above)
Comment (indietyp) [reply libs/@local/harpc/wire-protocol/src/codec/buffer.rs]: addressed in [05d2e73](https://github.com/hashintel/hash/pull/4389/commits/05d2e73b18cbc0207217af53062812175c631af9)

File: libs/@local/codec/src/harpc/wire.rs (hunk 1 above)
Comment (indietyp) [reply libs/@local/codec/src/harpc/wire.rs]: addressed in [05d2e73](https://github.com/hashintel/hash/pull/4389/commits/05d2e73b18cbc0207217af53062812175c631af9)

File: libs/@local/codec/src/harpc/wire.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+impl<T> Decoder for ProtocolDecoder<T>
+where
+    T: Decode<Context = ()>,
+{
+    type Error = Report<io::Error>;
+    type Item = T;
+
+    #[expect(clippy::big_endian_bytes, reason = "This is a protocol requirement")]
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/codec/src/harpc/wire.rs]: addressed in [05d2e73](https://github.com/hashintel/hash/pull/4389/commits/05d2e73b18cbc0207217af53062812175c631af9)

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**